text/0000-match-ergonmics.md

```diff
@@ -0,0 +1,531 @@
+- Feature Name: autoref_match
+- Start Date: 2016-08-12
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+Better ergonomics for match (and similar expressions).
+
+Prevent some of the boilerplate required for match expressions by auto-
+dereferencing the target expression, and 'auto-referencing' pattern variables.
+The rules for the latter are based closely on those for captured variables in
+closures.
+
+For example, current:
+
+```
+match *x {
+    Foo(ref x) => { ... }
+    Bar(ref mut y, z) => { ... }
+}
+```
+
+proposed:
+
+```
+match x {
+    Foo(x) => { ... }
+    Bar(y, z) => { ... }
+}
+```
+
+# Motivation
+[motivation]: #motivation
+
+Rust is usually strict when distinguishing between value and reference types. In
+particular, distinguishing borrowed and owned data. However, there is often a
+trade-off between [explicit-ness and ergonomics](https://blog.rust-lang.org/2017/03/02/lang-ergonomics.html),
+and Rust errs on the side of ergonomics in some carefully selected places.
+Notably when using the dot operator to call methods and access fields, and when
+declaring closures.
+
+The match expression is an extremely common expression and arguably, the most
+important control flow mechanism in Rust. Borrowed data is probably the most
+common form in the language. However, using match expressions and borrowed data
+together can be frustrating: getting the correct combination of `*`, `&`, and
+`ref` to satisfy the type and borrow checkers is a common problem, and one which
+is often encountered early by Rust beginners. It is especially frustrating since
+it seems that the compiler can guess what is needed but gives you error messages
+instead of helping.
+
+For example, consider the following program:
+
+```
+enum E { Foo(...), Bar }
+
+fn f(e: &E) {
+    match e { ... }
+}
+
+```
+
+It is clear what we want to do here - we want to check which variant `e` is a
+reference to. Annoyingly, we have two valid choices:
+
+```
+match e {
+    &E::Foo(...) => { ... }
+    &E::Bar => { ... }
+}
+```
+
+and
+
+```
+match *e {
+    E::Foo(...) => { ... }
+    E::Bar => { ... }
+}
+```
+
+The former is more obvious, but requires more noisey syntax (an `&` on every
+arm). The latter can appear a bit magical to newcomers - the type checker treats
+`*e` as a value, but the borrow checker treats the data as borrowed for the
+duration of the match. It also does not work with nested types, `match (*e,)
+...` for example is not allowed.
+
+In either case if we further bind variables, we must ensure that we do not
```

Comment:
This of course is still a problem if you don't use derefs - you'll just get "use of move data" errors instead of "move out of borrow" errors.